lib/Rabbit.js

```diff
@@ -0,0 +1,253 @@
+                const queueName = `${namespace}.${publisherName}.${serviceName}.${eventName}`;
+                const channel = await connection.createChannel();
+                await channel.assertQueue(queueName, {durable: true});
+                await channel.assertExchange(exchangeName, "fanout", {durable: true});
+                await channel.bindQueue(queueName, exchangeName, "");
+                channel.consume(queueName, (message) => {
+                    try {
+                        const payload = JSON.parse(message.content.toString());
+                        callback(payload);
+                        channel.ack(message);
+                    } catch (error) {
+                        channel.nack(message);
+                    }
+                });
+            }
+        }
+        // create notification channel & listen to event
+        for (const eventName in subNotification) {
+            const callback = subNotification[eventName];
+            const queueName = `${namespace}.${serviceName}.${eventName}`;
+            const channel = await connection.createChannel();
+            channel.assertQueue(queueName, {durable: true});
+            channel.consume(queueName, (message) => {
+                try {
+                    const payload = JSON.parse(message.content.toString());
+                    callback(payload);
+                    channel.ack(message);
+                } catch (error) {
+                    channel.nack(message);
+                }
+            });
+        }
+        return true;
+    }
+
+    async _pubQueue(connection) {
+        if (!this._config.pubQueue || !this._connection) {
+            return false;
+        }
+        // extract config
+        const { namespace, prefetch, pubQueue, pubPayload } = this._config;
+        // create channel & publish payload
+        for (const serviceName in pubQueue) {
+            const eventList = pubQueue[serviceName];
+            for (const eventName of eventList) {
+                const queueName = `${namespace}.queue.${serviceName}.${eventName}`;
+                // if queue is already used for publication, then skip it
+                if (this._publishedQueueList.includes(queueName)) {
+                    continue;
+                }
+                if ("pubPayload" in this._config) {
+                    const content = JSON.stringify(pubPayload);
+                    const channel = await connection.createChannel();
+                    await channel.prefetch(prefetch);
+                    channel.assertQueue(queueName, {durable: true});
+                    channel.sendToQueue(queueName, Buffer.from(content), { persistent: true });
+                    // note that the queue has already been used for publication
+                    this._publishedQueueList.push(queueName);
+                }
+            }
+        }
+        return true;
+    }
+
+    async _subQueue(connection) {
+        if (!this._config.subQueue) {
```

Comment:
I think we should add connection check on this line too

Reply:
suggestion accepted